Feature/#86 학습자료 crud 구현

lcqff opened the pull request:
## #️⃣ 연관된 이슈

close: #86

## 📝 작업 내용

학습자료 CRUD와 s3 파일(이미지가 아닌) 업로드를 구현했습니다. 

## 💬 리뷰 요구사항(선택)
- 넣을까 말까 긴가민가하면서 **MimeType 검사** 기능을 추가했습니다. [참고](https://isc9511.tistory.com/107) 사실 없어도 되는 기능이긴 한데 취약점이라 하길래 추가했어요🤔 필요없을거 같음 도로 빼겠습니다.
- 사용자가 중복된 파일(동일한 파일)을 업로드하는 경우에 대해서 따로 처리할 필요가 있나요?
- 컨트롤러의 CreateDocument의 RequestParts 이름이 **files**로 설정돼 있음에서 RestDocs 테스트에서 `partWithName("files").description("첨부 파일들")`와 같이 작성하면 file로 바꿔 적으라고 오류가 납니다....😬 이거 때문에 계속 찾아봤는데 해결 못하고 걍 올려용.......

lcqff committed:
feat: s3 파일 업로드 구현

- Mime 타입 판별

#87

lcqff committed:
chore: 의존성 추가

- tika: 자료의 Mime type 확인
- jsoup: 링크 내 이미지를 가져오기 위한 웹 크롤링

#86

lcqff committed:
Merge branch 'develop' of https://github.com/BDD-CLUB/01-doo-re-back into Feature/#86-학습자료_CRUD_구현

lcqff committed:
feat: 학습자료 CRUD 구현

#86

lcqff committed:
refactor: 스터디 삭제를 soft delete로 수정

#86

lcqff committed:
refactor: dto 수정

- DocumentCreateRequest: FileRequest 리스트 -> url String
- DocumentCondensedResponse, DocumentDetailResponse: member -> memberId

#86

lcqff committed:
fix: 파일과 Json 데이터를 동시에 전달하는 경우에는  RequestBody가 아닌 RequsetPart를 사용한다.

#86

lcqff committed:
test: Integretion 테스트, pojo 테스트 작성

#86

lcqff committed:
test: restdocs test 작성

#86

lcqff committed:
chore: 불필요한 코드 삭제

#86

lcqff commented on src/main/java/doore/file/application/S3DocumentFileService.java:
```diff
@@ -1,13 +1,26 @@
 @Component
 public class S3DocumentFileService extends S3FileService {
 
+    private static final List<String> DOCUMENT_FILE_EXTENSIONS = List.of(
+            "text/plain", "application/zip", //txt, zip
+            "application/pdf", "application/vnd.ms-powerpoint",
+            "application/vnd.openxmlformats-officedocument.presentationml.presentation", //pdf, ppt, pptx
+            "video/mp4", "video/x-msvideo", "video/webm", //mp4, avi, webm
+            "audio/mpeg", "audio/wav", "audio/webm"// mp3, wav, webm
+    );
```
업로드 가능한 파일이 어떤 것이 있는지 정하는 편이 좋을 것 같습니다!
일단은 뭐 가능한거 다 업로드 가능하도록 작성해뒀는데 어떤 걸 빼고 어떤 걸 추가할지 의견 받습니다!

lcqff replied:
그리고 이미지 학습자료가 아닌 파일 학습자료에도 이미지가 업로드 가능해야 할거 같은데 다른 분들 의견도 궁금합니다🤔 (이미지 학습자료처럼 미리보기가 뜨진 않고 첨부파일처럼 첨부될거같음)

shkisme replied:
> 그리고 이미지 학습자료가 아닌 파일 학습자료에도 이미지가 업로드 가능해야 할거 같은데 다른 분들 의견도 궁금합니다🤔 (이미지 학습자료처럼 미리보기가 뜨진 않고 첨부파일처럼 첨부될거같음)

넵 맞습니다~ 제가 S3DocumentFileService와 S3ImageFileService로 분리한 이유는, 학습 자료용 / 이미지용(팀 썸네일, 스터디 썸네일, 회원 썸네일 등...) 으로 나눈 것일 뿐이라서 학습자료에 이미지가 포함될 수 있다면 이미지 확장자들을 넣어주시면 될 것 같아용

lcqff replied:
아하 썸네일용으로 분리한거군요! 알겠습니다

lcqff committed:
chore: 불필요한 코드 삭제

#86

lcqff committed:
chore: shema 수정날짜 주석 추가

#86

lcqff commented on src/test/java/doore/restdocs/docs/DocumentDocsTest.java:
```diff
@@ -0,0 +1,157 @@
+                                parameterWithName("groupId").description("학습자료가 속한 그룹 id")
+                        ), requestParts(
+                                partWithName("request").description("학습자료 정보"),
+                                partWithName("file").description("첨부 파일들")
```
files로 작성하면 오류가 나는데 이유를 모르겠습니다😭 DocumentController의 RequestParts 인자명과 매치돼야 하는 거 아닌가요??

shkisme replied:
이건 getMockImageFile() 메서드가 뱉는 MockMultipartFile의 이름이 file로 지정되있어서 그런 것 같네요 😂

shkisme commented on build.gradle:
```diff
@@ -85,6 +85,10 @@ dependencies {
     asciidoctorExt 'org.springframework.restdocs:spring-restdocs-asciidoctor'
     testImplementation 'org.springframework.restdocs:spring-restdocs-mockmvc'
 
+    //file
+    compileOnly 'org.apache.tika:tika-parsers:1.20'
+    runtimeOnly 'org.apache.tika:tika-parsers:1.11'
```
이 의존성들은 왜 추가하셨는지 알 수 있을까요?

lcqff replied:
[MIME 타입 검사](https://isc9511.tistory.com/107) 용으로 추가했습니답

shkisme commented on src/main/java/doore/document/application/dto/response/DocumentCondensedResponse.java:
```diff
@@ -0,0 +1,13 @@
+public record DocumentCondensedResponse(
+        Long id,
+        String title,
+        String description,
+//        String thumbnailUrl,
```
주석이 있네요~

shkisme commented on src/main/java/doore/file/application/S3DocumentFileService.java:
```diff
@@ -1,13 +1,26 @@
 @Component
 public class S3DocumentFileService extends S3FileService {
 
+    private static final List<String> DOCUMENT_FILE_EXTENSIONS = List.of(
+            "text/plain", "application/zip", //txt, zip
+            "application/pdf", "application/vnd.ms-powerpoint",
+            "application/vnd.openxmlformats-officedocument.presentationml.presentation", //pdf, ppt, pptx
+            "video/mp4", "video/x-msvideo", "video/webm", //mp4, avi, webm
+            "audio/mpeg", "audio/wav", "audio/webm"// mp3, wav, webm
+    );
```
확장자는 파일의 `.` 뒤에 오는 이름으로 알고 있습니다!
ex) image.png라는 파일이 있다면 png가 확장자

뭔가 파일의 유형(?)을 리스트에 담아두신 것 같아서요~

lcqff replied:
MIME 타입을 담아뒀습니다!

02ggang9 replied:
변수 명을 수정해야 할 것 같습니다!

shkisme commented on src/main/java/doore/document/application/DocumentCommandService.java:
```diff
@@ -0,0 +1,141 @@
+                    .build();
+            fileRepository.save(newFile);
+            document.updateFiles(List.of(newFile));
+        } else {
```
else는 지양하는 건 어떨까요?
지난번에 else 로 인해서 버그 찾기 힘들었던 기억이 나네용

shkisme reviewed (commented):
수고하셨습니다~👍

shkisme commented on src/main/java/doore/document/domain/DocumentAccessType.java:
```diff
@@ -0,0 +1,5 @@
+package doore.document.domain;
+
+public enum DocumentAccessType {
+    teams, all,
```
Enum 선언은 대문자로 부탁드립니다

+) 후행쉼표도!

lcqff replied:
후행쉼표는 돼있는거 아닌가요??😯😯

lcqff replied:
앗 세미콜론 말씀하시는 건가 추가하겠습니다!

shkisme commented on src/main/java/doore/file/application/S3FileService.java:
```diff
@@ -22,17 +31,38 @@ public abstract class S3FileService {
+        return newFileName;
+    }
+
+    abstract FileExceptionType FailUploadErrorMessage();
```
메서드 명의 시작 글자는 소문자여야 합니다!

amaran-th commented on src/docs/asciidoc/document.adoc:
```diff
@@ -0,0 +1,73 @@
+include::{snippets}/document-get/path-parameters.adoc[]
+
+.Http Response
+include::{snippets}/document-get/http-response.adoc[]
```
response의 속성들에 대한 설명도 있으면 좋을 것 같아요

amaran-th commented on src/main/java/doore/document/application/DocumentCommandService.java:
```diff
@@ -0,0 +1,141 @@
+    private final S3DocumentFileService s3DocumentFileService;
+
+    public void createDocument(DocumentCreateRequest request, List<MultipartFile> multipartFiles,
+                               DocumentGroupType entityPath, Long entityId) {
```
[NIT]: entityPath, entityId보다는 좀더 직관적인 groupType, groupId는 어떤가요??

lcqff replied:
어머 이거 수정을 안했었네용

amaran-th commented on src/main/java/doore/document/application/DocumentCommandService.java:
```diff
@@ -0,0 +1,141 @@
+
+    private void validateDocumentType(DocumentType type, String url, List<MultipartFile> multipartFiles) {
+        if (type.equals(DocumentType.url) && url == null) {
+            throw new DocumentException(LINK_DOCUMENT_NEEDS_FILE);
```
LINK_DOCUMENT_NEEDS_URL이 더 적절해보입니다!

amaran-th commented on src/main/java/doore/document/domain/DocumentType.java:
```diff
@@ -0,0 +1,5 @@
+package doore.document.domain;
+
+public enum DocumentType {
+    image, document, url,
```
document는 엔티티의 이름이기도 하니 혼동하지 않도록 file이라고 표기하는 게 좋을 것 같아요!

amaran-th commented on src/main/java/doore/document/domain/DocumentGroupType.java:
```diff
@@ -0,0 +1,5 @@
+package doore.document.domain;
+
+public enum DocumentGroupType {
+    studies, teams,
```
Java에서 enum 원소는 관례적으로 대문자로 작성합니다!
[참고 글](https://inpa.tistory.com/entry/JAVA-%E2%98%95-%EC%97%B4%EA%B1%B0%ED%98%95Enum-%ED%83%80%EC%9E%85-%EB%AC%B8%EB%B2%95-%ED%99%9C%EC%9A%A9-%EC%A0%95%EB%A6%AC)

그리고 가급적이면 [단수형을 사용하는 것](https://hot-time.tistory.com/1663)이 좋다고 하네요~

amaran-th commented on src/main/java/doore/document/application/DocumentCommandService.java:
```diff
@@ -0,0 +1,141 @@
+            document.updateFiles(List.of(newFile));
+        } else {
+            List<String> filePaths = uploadFileToS3(document.getType(), multipartFiles);
+            List<File> newFiles = saveFile(filePaths, document);
```
여러 개의 파일을 한 번에 업로드하는 것이니 `uploadFilesToS3`,`saveFiles`로 복수형으로 표현해주는 게 좋을 것 같아요!

amaran-th commented on src/main/java/doore/document/application/DocumentCommandService.java:
```diff
@@ -0,0 +1,141 @@
+
+    public void deleteDocument(Long documentId) {
+        validateExistDocument(documentId);
+        documentRepository.deleteById(documentId);
```
Document(학습자료)를 삭제한다면 학습자료에 첨부된 파일들도 함께 삭제해야 하지 않을까요?

amaran-th commented on src/main/java/doore/document/application/DocumentQueryService.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    public DocumentDetailResponse getDocument(Long documentId) {
+        StudyDocument studyDocument = documentRepository.findById(documentId)
```
혹시 왜 StudyDocument라는 별도의 엔티티를 만드신건지 여쭤봐도 될까요?

lcqff replied:
StudyDocument == 저희가 만들었던 Document 엔티티입니다
라이브러리 클래스명 중에 Document라는 이름이 존재해서 헷갈릴거같아서 바꿨는데 돌려놓을까요...👀

amaran-th replied:
아하...그래도 Study 도메인과 연관이 있는 것으로 혼동이 올 수 있을 것 같아서 Document로 고치는 편이 좋아 보입니다!
사실 Member 도메인부터가 동일한 이름의 라이브러리 클래스가 있어서...ㅋㅋ

amaran-th commented on src/main/java/doore/study/domain/Study.java:
```diff
@@ -19,10 +19,12 @@
 
 @Entity
 @Getter
 @NoArgsConstructor(access = AccessLevel.PROTECTED)
```
`@SQLRestriction("is_deleted = false")`도 추가하면 어떨까요?

lcqff replied:
헉 너무 좋습니다👍

amaran-th reviewed (changes requested):
수고하셨습니다~~! 리뷰 확인 부탁드려요!

lcqff committed:
chore: 변수명 변경

- *_FILE_EXTENSIONS -> *_MIME_TYPES

#86

lcqff committed:
chore: 엔티티명 변경
- StudyDocument -> Document
#86

lcqff committed:
chore: 주석 제거
#86

lcqff committed:
refactor: paging 추가

#86

lcqff committed:
feat: 학습자료 업로드에 사진 확장자 추가

#86

lcqff committed:
chore: document rest docs의 requestPart 이름을 file에서 files로 변경

#86

lcqff committed:
refactor: ENUM은 대문자, 단수형으로 선언한다.

#86

lcqff committed:
refactor: else 를 사용하지 않는다.

#86

lcqff committed:
chore: enum 타입 끝에 세미콜론 추가

#86

lcqff committed:
chore: 대문자로 적힌 메서드 수정

#86

lcqff committed:
refactor: Enum DocumentType을 대문자로 수정

#86

lcqff committed:
refactor: DocumentDetailResponse에서 업로더 id 대신 업로더 이름을 반환하도록 수정

#86

lcqff committed:
chore: 파라미터명 변경, 커스텀예외명 변경

- entityType, entityId -> groupType, groupId
- LINK_DOCUMENT_NEEDS_FILE -> LINK_DOCUMENT_NEEDS_URL

#86

02ggang9 commented on src/main/java/doore/document/api/DocumentController.java:
```diff
@@ -0,0 +1,95 @@
+        if (groupType.equals("teams")) {
+            return DocumentGroupType.TEAM;
+        }
+        throw new DocumentException(NOT_FOUND_GROUP_TYPE);
```
enum에 필드 값을 추가하고 DocumentGroupType.values()로 매칭 되는지 확인한다면 코드가 조금 더 깔끔할 것 같습니다!

lcqff replied:
수정했는데 이걸 말씀하신게 맞겠죠?😅

02ggang9 commented on src/main/java/doore/document/api/DocumentController.java:
```diff
@@ -0,0 +1,95 @@
+        throw new DocumentException(NOT_FOUND_GROUP_TYPE);
+    }
+
+    @GetMapping(("/{documentId}"))
```
소괄호가 하나 더 추가된 것 같습니다!

lcqff replied:
어머나

02ggang9 commented on src/main/java/doore/document/application/DocumentCommandService.java:
```diff
@@ -0,0 +1,141 @@
+    }
+
+    private void validateExistGroup(DocumentGroupType groupType, Long groupId) {
+        if (groupType.equals(TEAM)) {
```
Enum 비교는 equals 보다 == 으로 비교하는 것이 좋을 것 같습니다!

이유는 아래와 같습니다!

https://johnmarc.tistory.com/152

lcqff replied:
저 혼란스럽습니다~~~~~~!!!
equals 메서드를 보면 (보내주신 링크에도 나와있 듯이) equals 메서드 내부에 이미 == 처리가 되어있는데  
```
public final boolean equals(Object other) {
        return this==other;
    }
```

어떤 점에서 ==이 equals보다 좋은 건지 글을 읽어봐도 잘 모르겠습니다🥺🥺🥺 ==가 NPE를 잡아주지 않는 점 때문에 == 사용을 권장하시는 건가요??

shkisme replied:
값을 비교하고자 할 때는 equals, 동일한 인스턴스인지 확인하고자 할 때는 == 를 쓰면 좋을 것 같다는 말씀인 것 같습니다!

lcqff replied:
equals를 사용해도 String이 아닌 경우에 한해서는 ==과 동일한 기능을 하는 거 같은데 단순 명시적 목적인가요??😯

02ggang9 commented on src/main/java/doore/document/application/DocumentCommandService.java:
```diff
@@ -0,0 +1,141 @@
+        List<String> urls = new ArrayList<>();
+        for (MultipartFile multipartFile : multipartFiles) {
+            String url = "";
+            if (type.equals(DocumentType.IMAGE)) {
```
이 부분도 마찬가지 입니다!

02ggang9 commented on src/main/java/doore/document/application/DocumentQueryService.java:
```diff
@@ -0,0 +1,62 @@
+            fileResponses.add(fileResponse);
+        }
+
+        return new DocumentDetailResponse(document.getId(), document.getName(), document.getDescription(),
```
앞의 PR가 마찬가지로 파라미터가 조금 많으면 빌더 사용도 좋을 것 같습니다~!

지금도 좋아요!!

02ggang9 commented on src/main/java/doore/file/application/S3FileService.java:
```diff
@@ -22,17 +31,38 @@ public abstract class S3FileService {
+        final String newFileName = createFileName(fileExtension);
+        final ObjectMetadata objectMetadata = getObjectMetadata(file);
+
+        try (final InputStream inputStream = file.getInputStream()) {
```
try-with-resource 사용 bb 입니다!

lcqff committed:
chore: ENUM DOCUMENT를 FILE로 수정

#86

lcqff committed:
chore: 파일 업로드 메서드를 복수형으로 변경

#86

lcqff committed:
refactor: DocumentGroupType에 String과 매치되는 GroupType 반환시키는 메서드를 추가한다.

#86

lcqff commented on src/test/java/doore/restdocs/docs/DocumentDocsTest.java:
```diff
@@ -0,0 +1,186 @@
+                ));
+    }
+
+    @Test
+    @Disabled //todo: Could not write JSON 오류 해결
+    @DisplayName("학습자료 목록을 조회한다.")
+    public void 학습자료_목록을_조회한다() throws Exception {
+        //given
+        DocumentCondensedResponse documentCondensedResponse =
+                new DocumentCondensedResponse(1L, "학습자료1", "학습자료1 입니다.", LocalDate.parse("2020-02-02"), 1L);
+        DocumentCondensedResponse otherDocumentCondensedResponse =
+                new DocumentCondensedResponse(2L, "학습자료2", "학습자료2 입니다.", LocalDate.parse("2020-02-03"), 2L);
+        Page<DocumentCondensedResponse> documentCondensedResponses = new PageImpl<>(
+                List.of(documentCondensedResponse, otherDocumentCondensedResponse));
+
+        //when
+        when(documentQueryService.getAllDocument(any(), any(), any(PageRequest.class)))
+                .thenReturn(documentCondensedResponses);
+
+        //then
+        final MultiValueMap<String, String> params = new LinkedMultiValueMap<>();
+        params.add("page", "0");
+        params.add("size", "4");
+        mockMvc.perform(get("/{groupType}/{groupId}/documents", "teams", 1).params(params))
+                .andExpect(status().isOk())
+                .andDo(document("document-get-list", pathParameters(
+                                parameterWithName("groupType").description("학습자료가 속한 그룹(teams/studies)"),
+                                parameterWithName("groupId").description("학습자료가 속한 그룹 id")
+                        ), queryParameters(
+                                parameterWithName("page").description("페이지 (default: 0)"),
+                                parameterWithName("size").description("한 페이지당 불러올 개수 (default: 4)")
+                        )
+                ));
+    }
```
헬프ㅠㅠ
Could not write Json 에러 떠서 실패하는데 이유를 못찾겠습니다...🥺
postman에선 잘돼요..

lcqff committed:
refactor: DocumentDetailResponse builder 변환
#86

JJimini commented on src/main/java/doore/document/api/DocumentController.java:
```diff
@@ -0,0 +1,85 @@
+    private final DocumentCommandService documentCommandService;
+    private final DocumentQueryService documentQueryService;
+
+    @PostMapping("/{groupType}/{groupId}/documents")
+    public ResponseEntity<Void> createDocument(@Valid @RequestPart DocumentCreateRequest request,
```
request에 붙어있는 RequestPart와 원래 저희가 사용하는 RequestBody의 차이가 궁금합니다~! RequestPart를 쓰신 이유가 있나요??

lcqff replied:
파일을 전달하는 경우 RequestPart 를 사용해야하는데 이때 DTO는 따로 ResquestBody로 전달하는게 불가능하더라구요! RequestPart로 같이 묶어서 전달하는 수밖에 없는거같습니다!

JJimini commented on src/main/java/doore/document/api/DocumentController.java:
```diff
@@ -0,0 +1,85 @@
+        DocumentGroupType group = DocumentGroupType.value(groupType);
+        Page<DocumentCondensedResponse> condensedDocuments =
+                documentQueryService.getAllDocument(group, groupId, PageRequest.of(page, size));
+        System.out.println(condensedDocuments.getContent());
+        return ResponseEntity.status(HttpStatus.OK).body(condensedDocuments);
```
...??? 왜...출력해주는 거죠...??

lcqff replied:
앗 문제있는 부분 테스트했던 코드인데 삭제가 안됐나봐요ㅠㅠ

JJimini commented on src/main/java/doore/document/application/DocumentQueryService.java:
```diff
@@ -0,0 +1,62 @@
+
+        return new DocumentDetailResponse(document.getId(), document.getName(), document.getDescription(),
+                document.getAccessType(), document.getType(), fileResponses,
+                document.getCreatedAt().toLocalDate(), "팜");
+    }
```
업로더는 항상 "팜" 인건가요??

lcqff replied:
헉!! 수정하겠습니다😂

JJimini commented on src/main/java/doore/document/domain/DocumentAccessType.java:
```diff
@@ -0,0 +1,6 @@
+
+public enum DocumentAccessType {
+    TEAM, ALL,
+    ;
```
개인적으로 enum 타입은 한줄에 하나씩 적는 것이 가독성이 더 좋은 것 같아요..!

JJimini commented on src/main/java/doore/document/domain/DocumentGroupType.java:
```diff
@@ -0,0 +1,21 @@
+
+public enum DocumentGroupType {
+    STUDY("studies"), TEAM("teams"),
+    ;
```
윗 코멘트와 같은 부분입니다!

JJimini reviewed (commented):
수고하셨습니다~! 테스트가 꼼꼼하시네요👍👍

lcqff committed:
chore: enum 개행
#86

lcqff committed:
chore: 충돌해결

#86

amaran-th reviewed (approved):
수고하셨습니다~!
